### PR TITLE
[RFC] Reset copyID also when dictionary is referenced

### LINKED ISSUE #5901: Possible memory leak in clear_tv
It appears that lists like `[&l [1], *l]` may produce a memory leak preserving until mark&sweep GC run (https://github.com/neovim/neovim/pull/5826#issuecomment-271038993). Need allocator logging from #5119 to construct a unit test, fixing should be trivial if I am right.

### PR DESCRIPTION
**Pull request by ZyX-I**
Fixes #5901.

### CODE REVIEW

File: src/nvim/memory.c [hunk 1]
```diff
@@ -488,13 +511,13 @@ void time_to_bytes(time_t time_, uint8_t buf[8])
 void free_all_mem(void)
 {
   buf_T       *buf, *nextbuf;
+  static bool entered = false;
 
-  // When we cause a crash here it is caught and Vim tries to exit cleanly.
-  // Don't try freeing everything again.
-  if (entered_free_all_mem) {
```
Comment (jamessan) [inline src/nvim/memory.c]: This is being checked in `func_unref`:

```
21653 void func_unref(char_u *name)
21654 {
21655   ufunc_T *fp;
21656  
21657   if (name != NULL && isdigit(*name)) {
21658     fp = find_func(name);
21659     if (fp == NULL) {
21660 #ifdef EXITFREE
21661       if (!entered_free_all_mem)  // NOLINT(readability/braces)
21662 #endif
21663         EMSG2(_(e_intern2), "func_unref()");
21664     } else {
21665       user_func_unref(fp);
21666     }
21667   }
21668 }
```

Comment (ZyX-I) [inline src/nvim/memory.c]: Wondering why this is not checked in #5119 where I copied the code from. Some new addition?

**Comment by ZyX-I** [comment]
Something new (from fifth job):

```
[ RUN      ] ...is/build/neovim/neovim/test/functional/core/job_spec.lua @ 328: jobs jobwait will run callbacks while waiting
...is/build/neovim/neovim/test/functional/core/job_spec.lua:346: Expected objects to be the same.
Passed in:
(nil)
Expected:
(table) {
  [1] = 'notification'
  [2] = 'wait'
  [3] = {
    [1] = 4 } }
stack traceback:
	...is/build/neovim/neovim/test/functional/core/job_spec.lua:346: in function <...is/build/neovim/neovim/test/functional/core/job_spec.lua:328>
[  FAILED  ] ...is/build/neovim/neovim/test/functional/core/job_spec.lua @ 328: jobs jobwait will run callbacks while waiting (232.26 ms)
==================== File /home/travis/build/neovim/neovim/build/log/ubsan.26109 ====================
= =================================================================
= ==26109==ERROR: AddressSanitizer: heap-use-after-free on address 0x61c00000e89c at pc 0x000000a1537b bp 0x7fff45011bd0 sp 0x7fff45011bc8
= READ of size 4 at 0x61c00000e89c thread T0
=     #0 0xa1537a in process_wait /home/travis/build/neovim/neovim/src/nvim/event/process.c:174:18
=     #1 0x8fd594 in f_jobwait /home/travis/build/neovim/neovim/src/nvim/eval.c:12498:18
=     #2 0x7b7630 in call_func /home/travis/build/neovim/neovim/src/nvim/eval.c:7285:11
=     #3 0x7ccdea in get_func_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:7090:11
=     #4 0x7c5dad in ex_call /home/travis/build/neovim/neovim/src/nvim/eval.c:2843:9
=     #5 0xb0af6a in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2198:5
=     #6 0xae9aa7 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
=     #7 0xad6573 in do_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2906:3
=     #8 0xad2a74 in cmd_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2644:14
=     #9 0xad2ba0 in ex_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2625:3
=     #10 0xb0af6a in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2198:5
=     #11 0xae9aa7 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
=     #12 0xaefdb5 in do_cmdline_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:273:10
=     #13 0x620873 in nvim_command /home/travis/build/neovim/neovim/src/nvim/api/vim.c:45:3
=     #14 0x568745 in handle_nvim_command /home/travis/build/neovim/neovim/build/src/nvim/auto/api/private/dispatch_wrappers.generated.h:1670:3
=     #15 0xfe69c3 in on_request_event /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/channel.c:455:19
=     #16 0xa09f9f in multiqueue_process_events /home/travis/build/neovim/neovim/src/nvim/event/multiqueue.c:146:7
=     #17 0x10bbd5f in nv_event /home/travis/build/neovim/neovim/src/nvim/normal.c:7894:3
=     #18 0x105e393 in normal_execute /home/travis/build/neovim/neovim/src/nvim/normal.c:1143:3
=     #19 0x16d3773 in state_enter /home/travis/build/neovim/neovim/src/nvim/state.c:58:26
=     #20 0x1013f3b in normal_enter /home/travis/build/neovim/neovim/src/nvim/normal.c:463:3
=     #21 0xdd38e2 in main /home/travis/build/neovim/neovim/src/nvim/main.c:542:3
=     #22 0x2ab60f661f44 in __libc_start_main /build/eglibc-oGUzwX/eglibc-2.19/csu/libc-start.c:287
=     #23 0x447005 in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0x447005)
= 
= 0x61c00000e89c is located 28 bytes inside of 1712-byte region [0x61c00000e880,0x61c00000ef30)
= freed by thread T0 here:
=     #0 0x4f1030 in __interceptor_free (/home/travis/build/neovim/neovim/build/bin/nvim+0x4f1030)
=     #1 0xefcf84 in xfree /home/travis/build/neovim/neovim/src/nvim/memory.c:130:3
=     #2 0x938b78 in free_term_job_data_event /home/travis/build/neovim/neovim/src/nvim/eval.c:23061:3
=     #3 0xa09f9f in multiqueue_process_events /home/travis/build/neovim/neovim/src/nvim/event/multiqueue.c:146:7
=     #4 0xa03c96 in loop_poll_events /home/travis/build/neovim/neovim/src/nvim/event/loop.c:56:3
=     #5 0xa15267 in process_wait /home/travis/build/neovim/neovim/src/nvim/event/process.c:173:5
=     #6 0x8fd594 in f_jobwait /home/travis/build/neovim/neovim/src/nvim/eval.c:12498:18
=     #7 0x7b7630 in call_func /home/travis/build/neovim/neovim/src/nvim/eval.c:7285:11
=     #8 0x7ccdea in get_func_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:7090:11
=     #9 0x7c5dad in ex_call /home/travis/build/neovim/neovim/src/nvim/eval.c:2843:9
=     #10 0xb0af6a in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2198:5
=     #11 0xae9aa7 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
=     #12 0xad6573 in do_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2906:3
=     #13 0xad2a74 in cmd_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2644:14
=     #14 0xad2ba0 in ex_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2625:3
=     #15 0xb0af6a in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2198:5
=     #16 0xae9aa7 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
=     #17 0xaefdb5 in do_cmdline_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:273:10
=     #18 0x620873 in nvim_command /home/travis/build/neovim/neovim/src/nvim/api/vim.c:45:3
=     #19 0x568745 in handle_nvim_command /home/travis/build/neovim/neovim/build/src/nvim/auto/api/private/dispatch_wrappers.generated.h:1670:3
=     #20 0xfe69c3 in on_request_event /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/channel.c:455:19
=     #21 0xa09f9f in multiqueue_process_events /home/travis/build/neovim/neovim/src/nvim/event/multiqueue.c:146:7
=     #22 0x10bbd5f in nv_event /home/travis/build/neovim/neovim/src/nvim/normal.c:7894:3
=     #23 0x105e393 in normal_execute /home/travis/build/neovim/neovim/src/nvim/normal.c:1143:3
=     #24 0x16d3773 in state_enter /home/travis/build/neovim/neovim/src/nvim/state.c:58:26
=     #25 0x1013f3b in normal_enter /home/travis/build/neovim/neovim/src/nvim/normal.c:463:3
=     #26 0xdd38e2 in main /home/travis/build/neovim/neovim/src/nvim/main.c:542:3
=     #27 0x2ab60f661f44 in __libc_start_main /build/eglibc-oGUzwX/eglibc-2.19/csu/libc-start.c:287
= 
= previously allocated by thread T0 here:
=     #0 0x4f14ec in calloc (/home/travis/build/neovim/neovim/build/bin/nvim+0x4f14ec)
=     #1 0xefd022 in xcalloc /home/travis/build/neovim/neovim/src/nvim/memory.c:144:15
=     #2 0x9307c8 in common_job_init /home/travis/build/neovim/neovim/src/nvim/eval.c:22959:27
=     #3 0x8c3a85 in f_jobstart /home/travis/build/neovim/neovim/src/nvim/eval.c:12384:27
=     #4 0x7b7630 in call_func /home/travis/build/neovim/neovim/src/nvim/eval.c:7285:11
=     #5 0x7ccdea in get_func_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:7090:11
=     #6 0x86a849 in eval7 /home/travis/build/neovim/neovim/src/nvim/eval.c:4353:15
=     #7 0x866235 in eval6 /home/travis/build/neovim/neovim/src/nvim/eval.c:4073:7
=     #8 0x860eb9 in eval5 /home/travis/build/neovim/neovim/src/nvim/eval.c:3925:7
=     #9 0x85c0c1 in eval4 /home/travis/build/neovim/neovim/src/nvim/eval.c:3666:7
=     #10 0x85b663 in eval3 /home/travis/build/neovim/neovim/src/nvim/eval.c:3588:7
=     #11 0x85ac03 in eval2 /home/travis/build/neovim/neovim/src/nvim/eval.c:3525:7
=     #12 0x7af7e2 in eval1 /home/travis/build/neovim/neovim/src/nvim/eval.c:3458:7
=     #13 0x8709b4 in get_list_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:4968:9
=     #14 0x86977d in eval7 /home/travis/build/neovim/neovim/src/nvim/eval.c:4292:21
=     #15 0x866235 in eval6 /home/travis/build/neovim/neovim/src/nvim/eval.c:4073:7
=     #16 0x860eb9 in eval5 /home/travis/build/neovim/neovim/src/nvim/eval.c:3925:7
=     #17 0x85c0c1 in eval4 /home/travis/build/neovim/neovim/src/nvim/eval.c:3666:7
=     #18 0x85b663 in eval3 /home/travis/build/neovim/neovim/src/nvim/eval.c:3588:7
=     #19 0x85ac03 in eval2 /home/travis/build/neovim/neovim/src/nvim/eval.c:3525:7
=     #20 0x7af7e2 in eval1 /home/travis/build/neovim/neovim/src/nvim/eval.c:3458:7
=     #21 0x7ccb16 in get_func_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:7076:9
=     #22 0x7c5dad in ex_call /home/travis/build/neovim/neovim/src/nvim/eval.c:2843:9
=     #23 0xb0af6a in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2198:5
=     #24 0xae9aa7 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
=     #25 0xad6573 in do_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2906:3
=     #26 0xad2a74 in cmd_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2644:14
=     #27 0xad2ba0 in ex_source /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:2625:3
=     #28 0xb0af6a in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2198:5
=     #29 0xae9aa7 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
= 
= SUMMARY: AddressSanitizer: heap-use-after-free /home/travis/build/neovim/neovim/src/nvim/event/process.c:174:18 in process_wait
= Shadow bytes around the buggy address:
=   0x0c387fff9cc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=   0x0c387fff9cd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=   0x0c387fff9ce0: 00 00 00 00 00 00 fa fa fa fa fa fa fa fa fa fa
=   0x0c387fff9cf0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=   0x0c387fff9d00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
= =>0x0c387fff9d10: fd fd fd[fd]fd fd fd fd fd fd fd fd fd fd fd fd
=   0x0c387fff9d20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=   0x0c387fff9d30: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=   0x0c387fff9d40: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=   0x0c387fff9d50: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=   0x0c387fff9d60: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
= Shadow byte legend (one shadow byte represents 8 application bytes):
=   Addressable:           00
=   Partially addressable: 01 02 03 04 05 06 07 
=   Heap left redzone:       fa
=   Heap right redzone:      fb
=   Freed heap region:       fd
=   Stack left redzone:      f1
=   Stack mid redzone:       f2
=   Stack right redzone:     f3
=   Stack partial redzone:   f4
=   Stack after return:      f5
=   Stack use after scope:   f8
=   Global redzone:          f9
=   Global init order:       f6
=   Poisoned by user:        f7
=   Container overflow:      fc
=   Array cookie:            ac
=   Intra object redzone:    bb
=   ASan internal:           fe
=   Left alloca redzone:     ca
=   Right alloca redzone:    cb
= ==26109==ABORTING
=====================================================================================================
```

Does not look related, yet something worse then regular random failure. Will restart the job though.

**Comment by ZyX-I** [comment]
No errors after restart. PR should be considered ready.

File: src/nvim/memory.c (hunk 1 above)
Comment (jamessan) [inline src/nvim/memory.c]: It was added recently, as a merge of Vim's 7.4.1868.  See 945540b7825e2909d17d8b8867330c3ace7c759c.

**Comment by justinmk** [comment]
ASAN failure happened again on master after merging this.

**Comment by oni-link** [comment]
@justinmk, cannot reproduce the ASAN failure, but does this patch fix the problem?

```diff
diff --git a/src/nvim/event/process.c b/src/nvim/event/process.c
index 39dd5fd55..dc7886469 100644
--- a/src/nvim/event/process.c
+++ b/src/nvim/event/process.c
@@ -170,8 +170,9 @@ int process_wait(Process *proc, int ms, MultiQueue *events)
   int status = -1;
   bool interrupted = false;
   if (!proc->refcount) {
+    status = proc->status;
     LOOP_PROCESS_EVENTS(proc->loop, proc->events, 0);
-    return proc->status;
+    return status;
   }
 
   if (!events) {
```

**Pull request merged**